### DISCUSSION TITLE
get the last 3 messages from a group

### DISCUSSION THREAD

**Discussion by MasIgor** [python-telegram-bot/python-telegram-bot#2977] [Q&A] [answered]
### Issue I am facing

Hello!

I am trying to get the last 3 messages sent to a group by any user except the bot itself. I have set the privacy accordingly to the bot, but I cannot figure out how to access old messages.. only the messages just sent..

Any help is greatly appreciated.

Thank you

### Operating System

Windows 10

### Version of Python, python-telegram-bot & dependencies

```shell
python-telegram-bot 13.11
Bot API 5.7
certifi 2021.10.08
Python 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 05:59:00) [MSC v.1929 64 bit (AMD64)]
```

**Comment by Poolitzer** [comment]
@tanzerlana Hey, the bot API in general has no methods about the history of anything, including messages in a chat. You can save every incoming message and delete the 4th one, but this will only work after the bot has joined the chat, not before.

You can do some hacky workaround by using the latest known message id of the chat, substracting one and copy/forwarding that message somewhere, but this can fail e.g. if its a service message or the message itself has been deleted.

**Comment by MasIgor** [comment]
@Poolitzer thank you for your answer. I try to explain what I am doing and why, so maybe there is a better way to do this.

We have machines that have a telegram bot running on them. its the same bot for all machines. these randomly send messages to a group chat about the machine status. it works great because we can throw the same exe on all machines, and each machine includes its ID in the message so we dont need to set up a new bot for each machine. now we would like to implement that we can send a config file to specific machines or to all machines. The problem is that if we send a message in chat, the first machine that reads it will evaluate it and the others will ignore it since it has "already been read" by the bot, so multiple instances of the same bot running on different machines will not read that message. we would be happy reading the last message from the chat. 3 would be better, but 1 is ok. it will be a file that is being sent, and based on the filename the machine will know if the file should be downloaded and saved or not. we want to avoid at all costs to make a bot for every single machine.

do you see some smart ways to solve this?

Thank you!

**Comment by Poolitzer** [comment, marked as answer]
> do you see some smart ways to solve this?

pin the message, check the `get_chat(chat_id).pinned_message` result every now and then? You can in that message reply to a message and you should get that one as well by doing `pinned_message.reply_to_message`. If it works (it should, just not checked), it would only work one level deep, so you get two messages instead of three, but maybe thats enough for you?

**Comment by MasIgor** [comment]
@Poolitzer  thank you very much.. so if I get it right, i would send the file, and pin it. that way I can check every few minutes if there is a pinned message, and would get that message regardless of when it was sent? is that correct? and I would have to save the ID of the message to avoid reading it multiple times.. right?

**Comment by Poolitzer** [comment]
> I can check every few minutes if there is a pinned message, and would get that message regardless of when it was sent?

jup

> and I would have to save the ID of the message to avoid reading it multiple times.. right?

I dont get that point. why would you want to read it multiple times?

**Comment by MasIgor** [comment]
If i execute get_chat(chat_id).pinned_message every 5 minutes, i will keep reading the same message over and over again. so by saving the message ID I would know if the pinned message changed since the last reading.. (unless i am missing something..)

**Reply by Poolitzer** [reply]
aha. Yes, sure, thats how I would implement it then as well. last_checked_message equals to the last message_id, and only if they differ go through the whole config stuff.

**Reply by MasIgor** [reply]
ok then all is clear! thank you very much @Poolitzer

**Reply by MasIgor** [reply]
I am trying it, but something must be wrong.. I tried it like this:

 message = self.bot.getChat(-xxxxxxxxxx).pinned_message
        print(message.text)

but I get the error AttributeError: 'bool' object has no attribute 'bot'

**Reply by MasIgor** [reply]
@Poolitzer  Ok, found the problem about that message, but pinned_message returns none.. even if thre is one pinned..
